### DISCUSSION TITLE
🧠 💡 Talk / Demo Idea Brainstorm

### DISCUSSION THREAD

**Discussion by sophiewigmore** [paketo-buildpacks/feedback#27] [Ideas]
Has a new and exciting feature come out in the project that you think others would be interested in? Have you had trouble with a specific use case, and could benefit from a demo? Have you discovered process pitfalls or triumphs in the project? 

We want to hear about it, and we want to share the knowledge! Demonstrating what Paketo Buildpacks can do, or explaining exciting new features with communities outside of our usual spheres of influence are great ways to engage with potential users/collaborators.

## Discussion
The goal of this discussion is to brain dump any and all conference talk and/or demo ideas people have. Whether it's an idea you intend on presenting yourself one day, or an idea for someone else to present that you think could be helpful, we'd love to hear about it here. Imagine it as a virtual whiteboard for us to get together and ideate! No idea is too big or too small.

**Comment by davidmirror-ops** [comment]
I'd like to see a workshop/hands on session where we walk through users on how to author a build pack from scratch.

BTW this is awesome, thank you for setting up this discussion!

**Reply by dmikusa** [reply]
I agree we need more education on this topic. The hard part with anything relating to buildpack authoring is that we are not consistent within the project about how we do this. There are two major libraries being used that overlap significantly but present different APIs. Any education efforts either have to touch on both, which is confusing/more work, or risk being wrong 50% of the time.

The other side to this is that because of the aforementioned split in libraries, we're trying to work to bring things back to one library but that's not a quick process. Change is likely though with both libraries through this process, so that makes education harder because you're learning stuff that might be legacy in short order.

**Reply by davidmirror-ops** [reply]
Got it, thanks Daniel :)

I'm wondering if this is clearly outlined in [the docs](https://paketo.io/docs/howto/create-paketo-buildpack/), because judging by them, Packit would be the only library.

**Reply by dmikusa** [reply]
No, it's not in the docs either which is not good either 😢 - Mostly because the part of the team using the undocumented library, Java buildpacks, hasn't really put a focus on authoring buildpacks so we just haven't really spent time on documenting how others can use it. That's something we are going to improve this year, but even then it's a challenge. Documenting two libraries and explaining how to pick between the two.

Not sure there's a good answer, but it may be an area we don't try to promote as much until we've done some work in this area. That's another thing on the 2022 roadmap.

**Reply by loewenstein** [reply]
Would it be a good start to write down why a separate approach was chosen although another one already existed? Or was this just a result of starting in parallel?

**Comment by sophiewigmore** [comment]
@fg-j and I put a previous talk proposal together previously about our experience with adding SBOMs to our builds, and the many iterations/learnings we've been through. We never presented this, but I still think it could be an interesting subject to talk about!


A demo in this area could be helpful too!

**Reply by davidmirror-ops** [reply]
Any chance you could put the proposal in max 1,000 characters?
There's a doc template ready for that: :)
https://docs.google.com/document/d/1vhl48mimk4H7sUftAqMFT1WTqyEoEsHzxemLgQtCVGM/edit?usp=sharing

The kind of talk most users (and content committees) seem to be interested in, have to do with 'experiences, learnings, etc' and how did you come through it, so definitely interesting to be submitted.

**Comment by sophiewigmore** [comment]
The .NET buildpacks have gone through somewhat of a makeover recently. If there are any .NET-focused opportunities, a workshop with the buildpacks or a talk explaining the capabilities could be good.

**Reply by fg-j** [reply]
Could be cool to demo .NET Buildpack + Steeltoe together, since there's already a Steeltoe community. We could get feedback from Steeltoe about what use-cases or demo apps would be appealing to show off.

**Reply by davidmirror-ops** [reply]
Seems like the next Steeltoe community meeting will happen in early July. 

Can I ping them to coordinate a demo within that timeframe?

**Reply by davidmirror-ops** [reply]
Could we have an abstract for this idea?
I pinged the .NET Munich Meetup leadership and they think it would be very beneficial for that community to learn more about the .NET Core Buildpack

This would be a virtual event (until further notice)

**Comment by ForestEckhardt** [comment]
Anatomy of a buildpack. A walkthrough of one of our buildpacks where we talk where we slowly peel back the API layers all the way down to the CNB spec and talk about the decisions that the buildpacks makes and how it conveys those decisions to the CNB lifecycle.

**Reply by davidmirror-ops** [reply]
This CNB talk at [Kubecon EU 2022](https://sched.co/ytrA) definitely reflects that this kind of 'under the hood' knowledge is needed.

Putting it into 1,000 characters max and elaborating on what the audience will gain from this session would be the next steps to make it happen :)

**Comment by ryanmoran** [comment]
Last year, Andy and I did a CF Summit workshop. The [content](https://github.com/cloudfoundry/summit-hands-on-labs/tree/master/2021/Paketo) could be updated, but it serves as a good walk-through of how to use buildpacks.

**Reply by davidmirror-ops** [reply]
This would be greatly beneficial for new users. Almost everything you need to submit this tutorial to a Linux Foundation event is right there in the repo :)

**Comment by dmikusa** [comment]
There are a number of Java-related features that folks don't seem to recognize exist/haven't found in our docs. The Tilt/reloadable process integration, the fantastic support we have around debugging/monitoring, and a few other bits (all things that make us stand out vs building with Dockerfiles). It would be nice to show them off more and improve awareness.

**Reply by davidmirror-ops** [reply]
[SpringOne CFP](https://springone.io/2022/cfp) closes June 20th. 

Besides localized meetups, this is an event where Paketo should definitely be in :)

**Reply by dmikusa** [reply]
David and I are planning to submit a topic. We've been workshopping it, but I can ping you with details.

**Reply by ramiyengar** [reply]
@davidmirror-ops , I've also been asking Chris C to consider sponsoring a booth at SpringOne this year. Especially because the Paketo Java game is so strong. Also, CF is a recognized brand at Spring conferences. The personas are a great fit.

**Comment by fg-j** [comment]
A workshop that demos buildpacks' value propositions beyond day 0 could be cool. Possibly for an audience of operators or devs at small orgs who also act as operators. Addressing questions like:
- How can I use SBOMs to identify when an image contains a vulnerability?
- How do I (quickly) rebase or rebuild at scale to address the vulnerability?
- How do I keep apps up to date with the latest versions of runtimes?

Could be a demo where "CVEs" come out during the talk and participants use buildpack features to get information/take action.

**Comment by fg-j** [comment]
An interactive talk that gets info from devs who currently use Dockerfiles to find out where they get their Dockerfiles, where they learned Dockerfile best practices, asks them how much they know about what's in the images that they build with Dockerfiles.

Compare with buildpacks, demonstrating that they produce images that are more optimized, more security hardened, do clever stuff with caching, etc.

This would also be an opportunity for the project to gain some information about the technology we're competing against.

**Comment by johnnyr0x** [comment]
DevOps at the edge. 

We could partner with users to highlight how they use Paketo to in their devOps process pushing changes in images to the edge (some places with low bandwidth). This highlights the value of buildpacks vs dockerfiles. 
-Smaller size
-Easier maintenance
-More secure (easier to address CVEs)

**Comment by sophiewigmore** [comment]
There may be some overlap with items here, but the core team did some ideation I wanted to share here:
Product-focused
1. (Talk or Demo) When to use Paketo Buildpacks instead of Dockerfiles
2. (Talk) DevOps on the Edge - using Paketo to scale (partner with a user or customer)
3. (Talk or Demo) What are buildpacks? How do they differ from standard ways to build an image?

Developers/Operators with Dockerfile Experience
1. (Talk) When to use Paketo Buildpacks instead of Dockerfiles 
2. (Demo or Workshop) How to move an app onto kpack/buildpacks from Dockerfiles. Inner feedback loop with pack then kpack
3. (Demo or Workshop) How to create + maintain a custom stack
4. (Talk) DevOps made easy using Paketo
5. (Talk) DevOps on the Edge - using Paketo to scale (partner with a user or customer)
6. (Demo or Workshop) You don’t need a platform (CF/Heroku/Tanzu) to use buildpacks)
7. (Demo or workshop) You can use buildpacks for local development

Someone with Buildpacks Experience/Buildpacks Authors
1. (Workshop) How can I create my own stack?
2. (Workshop) How to create + maintain a nontrivial custom buildpack (with packit)
3. (Demo) How to create your own buildpacks
4. (Demo) What is Occam and how does it work? How does it help?
5. (Demo) Can I build on top of the buildpack - use it as a library
6. (Demo) How to handle multiple processes (web+DB)
7. (Demo or Talk) What is the build plan, why is it important, and how does it work?
8. (Demo or Talk) SBOM/ and how to disable it in Node
9. (Talk) How you can contribute to Paketo

**Comment by ramiyengar** [comment]
Ship More, Script Less With Paketo Buildpacks

Here is a sample description from my most recent talk at Python Web Conf:

> Buildpacks, which appeared on the scene circa 2011, completely transformed the developer experience across the globe! First popularized by Heroku, they made single-click (or single-command) deployment a reality and appeared like magic. Several barriers to entry to using Heroku/Buildpacks - such as pricing, hosting, etc. - were solved by the use of Cloud Foundry, which prompted users to make use of a parallel standard for Buildpacks. With the mainstream adoption of containers, the build experience has transformed entirely. Consequently, the Buildpacks’ internals have also changed to allow the export of containers. After several years, the Buildpacks communities at Heroku and Cloud Foundry have converged over an open standard termed the Cloud Native Buildpacks specification. Implementations of this specification is available from Heroku, Paketo, Google, and several others.
> 
> In this talk, I aim to demonstrate the lifecycle and utility of Paketo Buildpacks, especially when using Python applications. By using Paketo (Cloud Native Buildpacks), this experience is greatly simplified for developers. The build-and-deploy phase is greatly simplified for all applications and services. Developers can essentially cut out all the noise that arises in the form of “config hell” while focusing solely on solving the core business problems.

I typically alter the language/framework to suit the conference. Outline of the actual talk is as follows: 

- Introduce the concept of Docker and containerization, highlighting the complexity of Dockerfiles when used extensively. 
- Introduce the idea of buildpacks through questions such as: What if you don't need Dockerfiles? What if there was the ability to "detect", "analyze", and "export" immutable artifacts? etc. 
- Clone a "random" repo from GitHub. 
- Run pack build and highlight the Paketo buildpacks. 
- Run the container and show the app executing. 
- Launch back into slides, introducing the CNB spec
- Buildpacks/Paketo layers. 
- Buildpacks/Paketo lifecycle
- Buildpacks/Paketo features
- Showcase of different PaaS-es using Buildpacks.

**Comment by ramiyengar** [comment]
Cloud Native CI/CD With Paketo Buildpacks & Tekton
This is a talk where I combine the build abilities of Paketo with the deployment workflows of Tekton pipelines. Copy of an abstract follows -

> The Kubernetes transformation is everywhere! In local dev, staging, deployments, monitoring, etc. This particular talk focuses on CI/CD pipelines that help with continuous delivery, when working with Kubernetes. Tekton is the de-facto tool for building pipelines with Kubernetes. A great way to spin up containers is to use Paketo Buildpacks. Combined, the two can provide a simple, clean, and secure way to setup CI/CD pipelines for teams that are interested in a cloud-native way to build and maintain deployment pipelines. Along with a technical demo, answers to several 'why's and 'how's about using Cloud Native Buildpacks/Paketo Buildpacks with Tekton will be given.

The talk is a demo where I basically step attendees through the steps available at https://buildpacks.io/docs/tools/tekton/
Prerequisites: Kubernetes cluster. 
Showcase: A half-dozen yaml files, k9s showing the clusters, featuring logs with "Paketo" in the output, etc.
Slide deck: Buildpacks, Paketo, Tekton pipeline visualization
